test/test-viz.cpp

```diff
@@ -0,0 +1,350 @@
+#define VER_SPACING 2
+
+struct Vertex {
+    std::string type;
+    std::string name;
+    int layer;
+    std::vector<Vertex> ins;
+    SDL_FRect position;
+};
+
+struct Line {
+    double x1;
+    double y1;
+    double x2;
+    double y2;
+};
+
+class logic_scheme {
+public:
+
+    logic_scheme() {
+        _outputNum = 0;
+    }
+
+    void add_input(const std::string& _name) {
+        Vertex input;
+        input.type = "IN";
+        input.name = _name;
+        input.layer = 0;
+
+        _scheme.emplace_back(input);
+    }
+
+    void add_output(const std::string& _input) {
+        Vertex output;
+        output.type = "OUT";
+        output.name = std::string("output" + std::to_string(_outputNum));
+
+        _outputNum++;
+
+        std::vector<std::string> ins;
+        ins.emplace_back(_input);
+
+        _queue.emplace_back(output, ins);
+
+        _scheme.emplace_back(output);
+    }
+
+    void add_dff(const std::string& _input, const std::string& _output) {
+        Vertex dff;
+
+        dff.type = "DFF";
+        dff.name = _output;
+
+        std::vector<std::string> ins;
+        ins.emplace_back(_input);
+
+        if(check_inputs(dff, ins)) _queue.emplace_back(dff, ins);
+
+        _scheme.emplace_back(dff);
+    }
+
+    void add_gate(const std::vector<std::string>& _inputs, const std::string& _output, const std::string& _type) {
+        Vertex gate;
+        gate.type = _type;
+        gate.name = _output;
+        std::vector<std::string> ins = _inputs;
+
+        if(check_inputs(gate, ins)) _queue.emplace_back(gate, ins);
+
+        _scheme.emplace_back(gate);
+    }
+
+    //Method to check whether demanded input exists or not
+    bool check_inputs(Vertex& element, std::vector<std::string>& ins) {
+        for(Vertex& suggested_element : _scheme) {
+            for(int i = 0; i < ins.size(); i++) {
+                if(ins[i] == suggested_element.name) {
+                    element.ins.emplace_back(suggested_element);
+                    element.layer = suggested_element.layer + 1;
+                    ins.erase(ins.begin() + i);
+                    i--;
+                }
+            }
+        }
+
+        if(ins.empty()) return false;
+        return true;
+    }
+
+    //Method to handle elements with lacking bonds
+    void request_handling() {
+        while(!_queue.empty()) {
+            for (int i = 0; i < _queue.size(); i++) {
+                //Going through every element of the scheme to find ones that are in the queue and connecting them to the inputs
+                for (int j = 0; j < _scheme.size(); j++) {
+                    if (std::get<0>(_queue[i]).name == _scheme[j].name) {
+                        if (!check_inputs(_scheme[j], std::get<1>(_queue[i]))) {
+                            _queue.erase(_queue.begin() + i);
+                            i--;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void output() {
```

Comment:
Providing a method to print to stdout is ok, but it is better to generalize a bit, e.g. pass an output stream as a method parameter.